[#45965] Setting definitions validations access nonloaded constants

ba1ash opened the pull request:
https://community.openproject.org/work_packages/45965
The core idea is to load settings lazily during the application initialization process. Then in `after_initialize` block load the rest of setting that are not used during the initialization. By doing so we avoid the original bug problem, because the validation of problematic setting values happens later.
To test that original problem has been resolved run the following against dev and this branch:
```
docker compose run --rm backend env OPENPROJECT_DEFAULT__PROJECTS__MODULES="- forums" bundle exec rails c
```

ba1ash committed:
[#45965] Setting definitions validations access nonloaded constants

https://community.openproject.org/work_packages/45965

ba1ash commented on config/constants/settings/definition.rb:
```diff
@@ -30,6 +30,996 @@ module Settings
+      },
+      default_projects_modules: {
+        default: %w[calendar board_view work_package_tracking news costs wiki],
+        allowed: -> { OpenProject::AccessControl.available_project_modules.map(&:to_s) }
```
@cbliard FYI Here and on line 349 of this file we have problematic validation back.

cbliard replied:
@ba1ash in which scenario?

cbliard replied:
Sorry I misunderstood yesterday. I thought you were saying that problem was back, but in fact you're saying that the original code to validate the value is back, and it is not causing any issues anymore. Good job!

cbliard commented on db/migrate/20190618115620_fix_available_languages.rb:
```diff
@@ -1,6 +1,6 @@
 class FixAvailableLanguages < ActiveRecord::Migration[5.2]
   def up
-    if Setting.exists?(name: 'available_languages')
+    if Setting.where(name: 'available_languages').exists?
```
```suggestion
    if Setting.where(name: 'available_languages').exists? # rubocop:ignore Rails/WhereExists
```

cbliard commented on db/migrate/20190618115620_fix_available_languages.rb:
```diff
@@ -16,7 +16,7 @@ def up
 
   def down
-    if Setting.exists?(name: 'available_languages')
+    if Setting.where(name: 'available_languages').exists?
```
```suggestion
    if Setting.where(name: 'available_languages').exists? # rubocop:ignore Rails/WhereExists
```

cbliard commented on db/migrate/20200903064009_enable_current_project_custom_fields_columns.rb:
```diff
@@ -1,6 +1,6 @@
 class EnableCurrentProjectCustomFieldsColumns < ActiveRecord::Migration[6.0]
   def up
-    return unless Setting.exists?(name: 'enabled_projects_column')
+    return unless Setting.where(name: 'enabled_projects_column').exists?
```
```suggestion
    return unless Setting.where(name: 'enabled_projects_column').exists? # rubocop:ignore Rails/WhereExists
```

cbliard commented on db/migrate/20211103120946_clean_emails_footer.rb:
```diff
@@ -1,6 +1,6 @@
 class CleanEmailsFooter < ActiveRecord::Migration[6.1]
   def up
-    return unless Setting.exists?(name: 'emails_footer')
+    return unless Setting.where(name: 'emails_footer').exists?
```
```suggestion
    return unless Setting.where(name: 'emails_footer').exists? # rubocop:ignore Rails/WhereExists
```

cbliard commented on db/migrate/20220811061024_rename_work_package_export_limit.rb:
```diff
@@ -1,6 +1,6 @@
 class RenameWorkPackageExportLimit < ActiveRecord::Migration[7.0]
   def up
-    if Setting.exists?(name: 'work_packages_projects_export_limit')
+    if Setting.where(name: 'work_packages_projects_export_limit').exists?
```
```suggestion
    if Setting.where(name: 'work_packages_projects_export_limit').exists? # rubocop:ignore Rails/WhereExists
```

ba1ash replied:
I think the right one would be: `# rubocop:disable Rails/WhereExists`. I used this version.
At least it silences Rubocop for me, but maybe I am missing something.

cbliard replied:
Yep sorry, I mixed it. I should have tested it before sending.

cbliard commented on lib_static/open_project/configuration.rb:
```diff
@@ -115,14 +115,6 @@ def define_config_methods(definition)
           TRUE_VALUES.include? self[definition.name]
         end
       end
-
-      # Filters a hash with String keys by a key prefix and removes the prefix from the keys
-      def settings_of_prefix(prefix)
-        Settings::Definition
-          .all_of_prefix(prefix)
-          .to_h { |setting| [setting.name.delete_prefix(prefix), setting.value] }
-          .symbolize_keys!
-      end
```
Thanks for cleaning dead code :+1:

cbliard commented on app/seeders/basic_data/setting_seeder.rb:
```diff
@@ -67,6 +67,7 @@ def data
 
     def seedable_setting_definitions
       Setting.definitions
+             .values
```
`Setting.definitions` and `Settings::Definition.all` both return the same thing and are used in the code. Maybe we could simplify code by removing one of the two methods.

For instance use `Setting.definitions` everywhere and keep `Settings::Definition.all` private
Or use only `Settings::Definition.all` and remove `Setting.definitions`.

What do you think?

ba1ash replied:
I thinks that's a good idea! I will remove `Settings.definitions`.

cbliard commented on config/application.rb:
```diff
@@ -214,8 +214,14 @@ class Application < Rails::Application
+    config.after_initialize do
+      Settings::Definition::DEFINITIONS.each do |setting_name, setting_options|
+        Settings::Definition.add(setting_name, **setting_options)
+      end
```
I would extract a `Settings::Definition.add_all` so that implementation details do not leak into `config/application.rb`.

cbliard commented on config/constants/settings/definition.rb:
```diff
@@ -152,9 +1142,10 @@ def add(name,
+        return if @all.present? && @all[name].present?
 
-        @by_name = nil
+        @all = {} if @all.nil?
```
That's minor but you can write it like `@all ||= {}`

You can even put this in the `#all` method so that you're guaranteed to always have a hash when accessing `#all`. This line would go away and all occurrences of `@all` can be replaced by `all`.

cbliard commented on config/constants/settings/definition.rb:
```diff
@@ -163,56 +1154,34 @@ def add(name,
-
-      def all_of_prefix(prefix)
-        all.select { |definition| definition.name.start_with?(prefix) }
+        @all.keys.include?(name.to_sym)
```
`#keys` will create an new array, which can be avoided by using method [`#key?`](https://ruby-doc.org/3.2.1/Hash.html#method-i-key-3F) instead (or `has_key?`, `member?`, or `include?` as they are all the same method).

cbliard replied:
Also I suspect this method will not work until at least one definition is loaded, as it would be `nil`. The problem would go away if using `all` with

```
def all
  @all ||= {}
end
```

ba1ash replied:
Great idea! Applied.

cbliard commented on config/constants/settings/definition.rb:
```diff
@@ -152,9 +1142,10 @@ def add(name,
               env_alias: nil)
-        return if @by_name.present? && @by_name[name.to_s].present?
+        name = name.to_sym
+        return if @all.present? && @all[name].present?
```
`#exists?` can be used here.

cbliard commented on config/constants/settings/definition.rb:
```diff
@@ -163,56 +1154,34 @@ def add(name,
+        @all[name] = definition
       end
 
       def define(&)
         instance_exec(&)
       end
```
I think this method is not needed anymore and can be removed.

ba1ash replied:
:+1:

cbliard reviewed (approved):
I like it. It works well and I see no problems with the code you have written.

I only have minor remarks that you are free to apply or reject. Feel free to merge whenever you want.

At some point I was suspecting that a setting added by a plugin through `Settings::Definition.add` and then reset in a test would not have a chance to be added anymore, as it does not belong to `DEFINITIONS`, but as the `@all` value is saved and restored, I suppose the impact is limited.

:+1: Thanks!

ba1ash committed:
Add comment to silence rubocop.

The reason for the complain to be not relevant is that we have our own
implementation of `Setting#exists?` which overrides the ActiveRecord's one.

ba1ash committed:
Remove `Setting.definitions`.

In favor for `Settings::Definition.all`.
Let's have one way for getting the list of core settings definitions.

ba1ash committed:
Refactor Settings::Definition.

- Add .add_all method. The purpose is to extract this logic from
  config/application.rb and have it in proper place and tested.
- Refactor .all method to initialize @all with {} if it is nil.
  It simlifies the usage significantly and makes code more stable.
